File: server/services/heartbeat.ts
/**
 * Interval-based heartbeat scheduler for agent check-ins.
 *
 * Creates a persistent Claude Code session per heartbeat check so the agent can
 * execute whatever HEARTBEAT.md instructs (check email, calendar, APIs, etc.).
 * When a heartbeat determines the user should be contacted, initiates an
 * outbound Twilio call and hands the live Claude session to the voice session
 * so it retains full context of what it checked.
 *
 * - Start/stop a 60-second global interval that checks all enabled agents
 * - Track per-agent check intervals and concurrent-check guards
 * - Create persistent Claude sessions with full tool access
 * - Parse JSON heartbeat responses and initiate outbound calls
 * - Pass live Claude sessions to the Twilio server for voice call continuity
 * - Expose last heartbeat results for the API
 */

import { randomUUID } from "crypto";
import { readFileSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

import twilio from "twilio";
import { createClaudeSession, type ClaudeSession } from "../voice/claude-session.js";
import { listAgents, getAgent, AGENTS_DIR, type Agent } from "./agent-store.js";
import { readEnv } from "./env.js";
import { getTunnelUrl, isTunnelRunning } from "./tunnel.js";
import { isRunning as isTwilioRunning } from "./twilio-manager.js";
import { setCallClaudeSession } from "../voice/twilio-server.js";

// ============================================================================
// CONSTANTS
// ============================================================================

const __dirname = dirname(fileURLToPath(import.meta.url));

/** Global check interval in milliseconds (60 seconds) */
const CHECK_INTERVAL_MS = 60_000;

/** Maximum time for a single heartbeat Claude session in milliseconds */
const SESSION_TIMEOUT_MS = 120_000;

/** User-facing prompt sent to the heartbeat Claude session */
const HEARTBEAT_PROMPT = readFileSync(join(__dirname, "..", "..", "init", "defaults", "system-heartbeat.md"), "utf-8").trim();

/** Default voice system prompt (shared with voice sessions) */
const DEFAULT_SYSTEM_PROMPT = readFileSync(join(__dirname, "..", "..", "init", "defaults", "system.md"), "utf-8").trim();

// ============================================================================
// TYPES
// ============================================================================

/** Result of a single agent heartbeat check */
export interface HeartbeatResult {
  agentId: string;
  shouldCall: boolean;
  reason: string;
  timestamp: number;
}

// ============================================================================
// STATE
// ============================================================================

/** Global setInterval handle */
let intervalTimer: ReturnType<typeof setInterval> | null = null;

/** Last heartbeat result per agent */
let lastResults: Record<string, HeartbeatResult> = {};

/** Last check timestamp per agent (for interval tracking) */
let lastCheckTimes: Record<string, number> = {};

/** Currently running agent IDs (concurrent guard) */
const inFlightChecks = new Set<string>();

// ============================================================================
// MAIN HANDLERS
// ============================================================================

/**
 * Start the heartbeat scheduler.
 * Runs checkAllAgents every 60 seconds via setInterval.
 */
export function startHeartbeat(): void {
  if (intervalTimer) return;

  intervalTimer = setInterval(() => {
    checkAllAgents().catch((err) => {
      console.error("[heartbeat] checkAllAgents error:", err);
    });
  }, CHECK_INTERVAL_MS);

  console.log("[heartbeat] scheduler started (60s interval)");
}

/**
 * Stop the heartbeat scheduler.
 * Clears the global interval timer.
 */
export function stopHeartbeat(): void {
  if (intervalTimer) {
    clearInterval(intervalTimer);
    intervalTimer = null;
  }
}

/**
 * Get the last heartbeat result per agent.
 * Used by the API to expose heartbeat status.
 *
 * @returns Record of agent ID to last HeartbeatResult
 */
export function getHeartbeatStatus(): Record<string, HeartbeatResult> {
  return lastResults;
}

/**
 * Initiate an outbound Twilio call to an agent's phone number.
 * Used by both the heartbeat scheduler and the API "Call Me" route.
 *
 * Flow:
 * 1. Check preconditions (tunnel + Twilio server running)
 * 2. Generate UUID token and register it with the Twilio server
 * 3. Optionally attach a live Claude session for voice call continuity
 * 4. Place outbound call via Twilio SDK with TwiML streaming to our WebSocket
 *
 * @param agent - Full agent data including config with phone number
 * @param claudeSession - Optional live Claude session to hand off to the voice call
 * @returns The Twilio call SID
 */
export async function initiateAgentCall(agent: Agent, opts?: { claudeSession?: ClaudeSession; initialPrompt?: string }): Promise<string> {
  // Check preconditions
  if (!isTunnelRunning()) {
    throw new Error("Tunnel is not running. Cannot place outbound call.");
  }
  if (!isTwilioRunning()) {
    throw new Error("Twilio server is not running. Cannot place outbound call.");
  }

  const token = randomUUID();
  const envVars = await readEnv();

  const twilioPort = parseInt(envVars.TWILIO_PORT || "8080", 10);
  const accountSid = envVars.TWILIO_ACCOUNT_SID;
  const authToken = envVars.TWILIO_AUTH_TOKEN;
  const userPhoneNumber = envVars.USER_PHONE_NUMBER;

  if (!accountSid || !authToken) {
    throw new Error("TWILIO_ACCOUNT_SID and TWILIO_AUTH_TOKEN must be set in .env");
  }
  if (!userPhoneNumber) {
    throw new Error("USER_PHONE_NUMBER must be set in Settings > General");
  }

  // Register the call token with the Twilio server (with optional initial prompt)
  await registerCallToken(twilioPort, token, agent.id, opts?.initialPrompt);

  // Attach the live Claude session if provided (heartbeat-initiated calls)
  if (opts?.claudeSession) {
    setCallClaudeSession(token, opts.claudeSession);
  }

  // Get the tunnel URL (strip protocol for WebSocket URL)
  const fullTunnelUrl = getTunnelUrl()!;
  const tunnelHost = fullTunnelUrl.replace(/^https?:\/\//, "");

  // Get the first Twilio phone number on the account
  const client = twilio(accountSid, authToken);
  const numbers = await client.incomingPhoneNumbers.list({ limit: 1 });
  if (numbers.length === 0) {
    throw new Error("No Twilio phone numbers found on the account");
  }
  const fromNumber = numbers[0].phoneNumber;

  // Build TwiML with WebSocket stream
  const twiml = `<Response><Connect><Stream url="wss://${tunnelHost}/media/${token}?agentId=${agent.id}" /></Connect></Response>`;

  // Place the outbound call
  const call = await client.calls.create({
    to: userPhoneNumber,
    from: fromNumber,
    twiml,
  });

  console.log(`[heartbeat] outbound call placed to ${userPhoneNumber} (callSid=${call.sid})`);
  return call.sid;
}

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Check all enabled agents and spawn heartbeat sessions for those that are due.
 * Skips agents whose configured interval has not elapsed and agents with
 * in-flight checks (concurrent guard).
 */
async function checkAllAgents(): Promise<void> {
  const summaries = await listAgents();
  const enabledSummaries = summaries.filter((s) => s.enabled);

  if (enabledSummaries.length === 0) return;

  const now = Date.now();

  for (const summary of enabledSummaries) {
    // Skip if interval has not elapsed
    const lastCheck = lastCheckTimes[summary.id] ?? 0;
    const intervalMs = summary.heartbeatIntervalMinutes * 60_000;
    if (now - lastCheck < intervalMs) continue;

    // Skip if already checking this agent
    if (inFlightChecks.has(summary.id)) continue;

    // Load full agent data and spawn check (fire-and-forget)
    const agent = await getAgent(summary.id);
    checkSingleAgent(agent).catch((err) => {
      console.error(`[heartbeat] check failed for agent "${agent.id}":`, err);
    });
  }
}

/**
 * Run a heartbeat check for a single agent using a persistent Claude session.
 * If the check determines shouldCall, keeps the session alive and passes it
 * to the outbound call so the voice session continues with full context.
 *
 * @param agent - Full agent data with SOUL.md, MEMORY.md, HEARTBEAT.md
 */
async function checkSingleAgent(agent: Agent): Promise<HeartbeatResult> {
  inFlightChecks.add(agent.id);
  lastCheckTimes[agent.id] = Date.now();

  let session: ClaudeSession | null = null;

  try {
    const { result, claudeSession } = await runHeartbeatSession(agent);
    session = claudeSession;
    lastResults[agent.id] = result;

    console.log(
      `[heartbeat] agent "${agent.id}": shouldCall=${result.shouldCall}, reason="${result.reason}"`,
    );

    if (result.shouldCall) {
      try {
        // Pass the live session to the call — it will be handed to the voice session
        await initiateAgentCall(agent, { claudeSession: session });
        session = null; // Don't close — voice session owns it now
      } catch (err) {
        console.error(`[heartbeat] failed to call agent "${agent.id}":`, err);
      }
    }

    return result;
  } finally {
    // Close the session if we still own it (shouldCall was false, or call failed)
    if (session) {
      await session.close();
    }
    inFlightChecks.delete(agent.id);
  }
}

/**
 * Run a heartbeat check using a persistent Claude session.
 * Creates the session with the agent's full context (voice instructions +
 * SOUL.md + MEMORY.md + HEARTBEAT.md), sends the heartbeat prompt, and
 * parses the JSON response.
 *
 * Returns both the parsed result and the live session so the caller can
 * decide whether to keep it alive for a voice call.
 *
 * @param agent - Full agent data
 * @returns The heartbeat result and the live Claude session
 */
async function runHeartbeatSession(agent: Agent): Promise<{ result: HeartbeatResult; claudeSession: ClaudeSession }> {
  // Include voice instructions so the session is ready for voice call continuity
  const systemPrompt = [DEFAULT_SYSTEM_PROMPT, agent.soulMd, agent.memoryMd, agent.heartbeatMd].join("\n\n");

  const claudeSession = await createClaudeSession({
    allowedTools: [],
    permissionMode: "bypassPermissions",
    systemPrompt: "",
    customSystemPrompt: systemPrompt,
    cwd: join(AGENTS_DIR, agent.id),
  });

  // Set up a timeout to close the session if it takes too long
  let timedOut = false;
  const timeout = setTimeout(() => {
    timedOut = true;
    claudeSession.interrupt();
  }, SESSION_TIMEOUT_MS);

  try {
    // Send the heartbeat prompt and collect the response text
    let responseText = "";
    const eventStream = claudeSession.sendMessage(HEARTBEAT_PROMPT);

    for await (const event of eventStream) {
      if (event.type === "text_delta") {
        responseText += event.content;
      }
      if (event.type === "result") break;
    }

    if (timedOut) {
      console.error(`[heartbeat] session timed out for agent "${agent.id}"`);
      return { result: failSafeResult(agent.id), claudeSession };
    }

    if (!responseText) {
      console.error(`[heartbeat] no response text for agent "${agent.id}"`);
      return { result: failSafeResult(agent.id), claudeSession };
    }

    const result = parseHeartbeatResponse(agent.id, responseText);
    return { result, claudeSession };
  } catch (err) {
    if (timedOut) {
      console.error(`[heartbeat] session timed out for agent "${agent.id}"`);
    } else {
      console.error(`[heartbeat] session error for agent "${agent.id}":`, err);
    }
    return { result: failSafeResult(agent.id), claudeSession };
  } finally {
    clearTimeout(timeout);
  }
}

/**
 * Parse a heartbeat JSON response string into a HeartbeatResult.
 * Expects a JSON object with shouldCall (boolean) and reason (string).
 * Returns a fail-safe result if parsing fails.
 *
 * @param agentId - Agent identifier for the result
 * @param text - Raw text from the assistant response
 * @returns Parsed HeartbeatResult
 */
function parseHeartbeatResponse(agentId: string, text: string): HeartbeatResult {
  try {
    // Extract JSON from the text (may contain surrounding text)
    const jsonMatch = text.match(/\{[\s\S]*"shouldCall"[\s\S]*\}/);
    if (!jsonMatch) {
      console.error(`[heartbeat] no JSON found in response for agent "${agentId}": ${text}`);
      return failSafeResult(agentId);
    }

    const parsed = JSON.parse(jsonMatch[0]);
    return {
      agentId,
      shouldCall: Boolean(parsed.shouldCall),
      reason: String(parsed.reason || ""),
      timestamp: Date.now(),
    };
  } catch (err) {
    console.error(`[heartbeat] JSON parse error for agent "${agentId}":`, err);
    return failSafeResult(agentId);
  }
}

/**
 * Create a fail-safe HeartbeatResult that does not trigger a call.
 * Used when the session errors, times out, or returns unparseable output.
 *
 * @param agentId - Agent identifier
 * @returns HeartbeatResult with shouldCall: false
 */
function failSafeResult(agentId: string): HeartbeatResult {
  return {
    agentId,
    shouldCall: false,
    reason: "heartbeat check failed or timed out",
    timestamp: Date.now(),
  };
}

/**
 * Register a call token with the Twilio server via HTTP POST.
 * Even though the Twilio server now runs in-process, we still use the HTTP
 * endpoint to register tokens since the WebSocket upgrade path validates
 * against the activeCalls map populated by this endpoint.
 *
 * @param port - Twilio server port
 * @param token - UUID token for the call
 * @param agentId - Agent identifier to associate with the call
 */
async function registerCallToken(port: number, token: string, agentId: string, initialPrompt?: string): Promise<void> {
  const body = JSON.stringify({ token, agentId, ...(initialPrompt && { initialPrompt }) });

  const response = await fetch(`http://localhost:${port}/register-call`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body,
  });

  if (!response.ok) {
    throw new Error(`Failed to register call token: ${response.status} ${response.statusText}`);
  }
}

File: voice-server/heartbeat.py
"""
Heartbeat scheduler for agent check-ins, ported from heartbeat.ts.

Creates a persistent Claude Code session per heartbeat check so the agent can
execute whatever HEARTBEAT.md instructs. When a heartbeat determines the user
should be contacted, initiates an outbound Twilio call and hands the live Claude
session to the voice pipeline so it retains full context.

Responsibilities:
- Start/stop a 60-second asyncio interval that checks all enabled agents
- Track per-agent check intervals and concurrent-check guards
- Create persistent Claude sessions with full tool access
- Parse JSON heartbeat responses and initiate outbound calls
- Store live Claude sessions in pending_calls for voice call handoff
- 60-second cleanup timer for unanswered calls
- Expose last heartbeat results for the API
"""

import asyncio
import json
import logging
import os
import re
import time
from dataclasses import dataclass, field
from uuid import uuid4

from claude_agent_sdk import ClaudeAgentOptions, ClaudeSDKClient, AssistantMessage, ResultMessage, TextBlock

from config import (
    Agent,
    VoiceServerConfig,
    build_system_prompt,
    list_agents,
    load_agent,
    DEFAULT_AGENTS_DIR,
    PROJECT_ROOT,
)

logger = logging.getLogger(__name__)

# ============================================================================
# CONSTANTS
# ============================================================================

# Global check interval (60 seconds)
CHECK_INTERVAL_S = 60

# Default max time for a single heartbeat Claude session (5 minutes)
DEFAULT_HEARTBEAT_TIMEOUT_S = 5 * 60

# Cleanup timer for unanswered pending calls (60 seconds)
PENDING_CALL_TIMEOUT_S = 60

# Heartbeat prompt sent to the Claude session
_HEARTBEAT_PROMPT_PATH = os.path.join(PROJECT_ROOT, "init", "defaults", "system-heartbeat.md")


def _load_heartbeat_prompt() -> str:
    """Read the heartbeat prompt from disk."""
    with open(_HEARTBEAT_PROMPT_PATH, "r", encoding="utf-8") as f:
        return f.read().strip()


# ============================================================================
# TYPES
# ============================================================================

@dataclass
class HeartbeatResult:
    """Result of a single agent heartbeat check."""
    agent_id: str
    should_call: bool
    reason: str
    timestamp: float


@dataclass
class PendingCall:
    """A pending outbound call waiting for Twilio to connect."""
    token: str
    agent_id: str
    client: ClaudeSDKClient
    initial_prompt: str | None
    created_at: float


# ============================================================================
# STATE
# ============================================================================

# Asyncio task for the interval loop
_interval_task: asyncio.Task | None = None

# Last heartbeat result per agent
_last_results: dict[str, HeartbeatResult] = {}

# Last check timestamp per agent (for interval tracking)
_last_check_times: dict[str, float] = {}

# Currently running agent IDs (concurrent guard)
_in_flight_checks: set[str] = set()

# Pending calls keyed by token, waiting for Twilio WebSocket connection
_pending_calls: dict[str, PendingCall] = {}

# Reference to config (set on start)
_config: VoiceServerConfig | None = None

# Getter for tunnel URL (set on start, imported from server module)
_get_tunnel_url = None


# ============================================================================
# MAIN HANDLERS
# ============================================================================

def start_heartbeat(config: VoiceServerConfig, get_tunnel_url_fn) -> None:
    """Start the heartbeat scheduler.

    Runs _check_all_agents every 60 seconds via an asyncio task.

    Args:
        config: Voice server configuration with Twilio credentials
        get_tunnel_url_fn: Callable that returns the current tunnel URL
    """
    global _interval_task, _config, _get_tunnel_url

    if _interval_task is not None:
        return

    _config = config
    _get_tunnel_url = get_tunnel_url_fn

    _interval_task = asyncio.create_task(_interval_loop())
    logger.info("[heartbeat] scheduler started (60s interval)")


def stop_heartbeat() -> None:
    """Stop the heartbeat scheduler. Cancels the asyncio interval task."""
    global _interval_task

    if _interval_task is not None:
        _interval_task.cancel()
        _interval_task = None
        logger.info("[heartbeat] scheduler stopped")


def get_heartbeat_status() -> dict[str, dict]:
    """Get the last heartbeat result per agent.

    Returns:
        Dict of agent_id -> serialized HeartbeatResult
    """
    return {
        agent_id: {
            "agent_id": r.agent_id,
            "should_call": r.should_call,
            "reason": r.reason,
            "timestamp": r.timestamp,
        }
        for agent_id, r in _last_results.items()
    }


def get_pending_client(token: str) -> PendingCall | None:
    """Retrieve and remove a pending call by token.

    Called when the Twilio WebSocket connects to hand off the live Claude session.

    Args:
        token: The call token

    Returns:
        PendingCall if found, None otherwise
    """
    return _pending_calls.pop(token, None)


def register_pending_call(token: str, agent_id: str, initial_prompt: str | None = None) -> None:
    """Register a pending call without a pre-existing Claude client.

    Used by the /register-call endpoint for API-initiated calls (e.g. "Call Me").

    Args:
        token: Unique call token
        agent_id: Agent identifier
        initial_prompt: Optional initial prompt for the agent
    """
    # No Claude client -- the pipeline will create a fresh one
    _pending_calls[token] = PendingCall(
        token=token,
        agent_id=agent_id,
        client=None,  # type: ignore[arg-type]
        initial_prompt=initial_prompt,
        created_at=time.time(),
    )


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

async def _interval_loop() -> None:
    """Asyncio loop that runs _check_all_agents every CHECK_INTERVAL_S."""
    while True:
        try:
            await _check_all_agents()
        except Exception as e:
            logger.error(f"[heartbeat] check_all_agents error: {e}")
        await asyncio.sleep(CHECK_INTERVAL_S)


async def _check_all_agents() -> None:
    """Check all enabled agents and spawn heartbeat sessions for those that are due."""
    if not _config:
        return

    agents = list_agents(_config.agents_dir)
    if not agents:
        return

    now = time.time()

    for agent in agents:
        # Skip if interval has not elapsed
        last_check = _last_check_times.get(agent.id, 0)
        interval_s = agent.config.heartbeat_interval_minutes * 60
        if now - last_check < interval_s:
            continue

        # Skip if already checking this agent
        if agent.id in _in_flight_checks:
            continue

        # Fire-and-forget the check
        asyncio.create_task(_check_single_agent_wrapper(agent))


async def _check_single_agent_wrapper(agent: Agent) -> None:
    """Wrapper for check_single_agent that handles errors and in-flight tracking."""
    try:
        await check_single_agent(agent)
    except Exception as e:
        logger.error(f'[heartbeat] check failed for agent "{agent.id}": {e}')


async def check_single_agent(agent: Agent) -> HeartbeatResult:
    """Run a heartbeat check for a single agent using a persistent Claude session.

    If the check determines shouldCall, keeps the session alive and passes it
    to the outbound call so the voice session continues with full context.

    Args:
        agent: Full agent data with SOUL.md, MEMORY.md, HEARTBEAT.md

    Returns:
        HeartbeatResult with the check outcome
    """
    _in_flight_checks.add(agent.id)
    _last_check_times[agent.id] = time.time()

    client: ClaudeSDKClient | None = None

    try:
        timeout_s = (agent.config.heartbeat_timeout_minutes or 5) * 60 or DEFAULT_HEARTBEAT_TIMEOUT_S
        result, client = await _run_heartbeat_session(agent, timeout_s)
        _last_results[agent.id] = result

        logger.info(
            f'[heartbeat] agent "{agent.id}": shouldCall={result.should_call}, '
            f'reason="{result.reason}"'
        )

        if result.should_call:
            try:
                await initiate_agent_call(agent, client)
                client = None  # Don't close -- voice session owns it now
            except Exception as e:
                logger.error(f'[heartbeat] failed to call agent "{agent.id}": {e}')

        return result
    finally:
        # Close the session if we still own it
        if client:
            try:
                await client.disconnect()
            except Exception:
                pass
        _in_flight_checks.discard(agent.id)


async def _run_heartbeat_session(
    agent: Agent, timeout_s: float
) -> tuple[HeartbeatResult, ClaudeSDKClient]:
    """Run a heartbeat check using a persistent Claude session.

    Creates the session with the agent's full context, sends the heartbeat prompt,
    and parses the JSON response.

    Args:
        agent: Full agent data
        timeout_s: Maximum time for the session in seconds

    Returns:
        Tuple of (HeartbeatResult, live ClaudeSDKClient)
    """
    agent_dir = os.path.join(DEFAULT_AGENTS_DIR, agent.id)
    system_prompt = build_system_prompt(agent.id, "voice")

    options = ClaudeAgentOptions(
        system_prompt=system_prompt,
        cwd=agent_dir,
        allowed_tools=[],
        permission_mode="bypassPermissions",
        include_partial_messages=True,
        max_thinking_tokens=10000,
    )
    client = ClaudeSDKClient(options=options)
    await client.connect()

    timed_out = False

    async def _timeout_guard():
        nonlocal timed_out
        await asyncio.sleep(timeout_s)
        timed_out = True
        try:
            await client.interrupt()
        except Exception:
            pass

    timeout_task = asyncio.create_task(_timeout_guard())

    try:
        heartbeat_prompt = _load_heartbeat_prompt()
        response_text = ""

        await client.query(heartbeat_prompt)

        async for msg in client.receive_response():
            if isinstance(msg, AssistantMessage):
                for block in msg.content:
                    if isinstance(block, TextBlock) and block.text:
                        response_text += block.text
            elif isinstance(msg, ResultMessage):
                break

        if timed_out:
            logger.error(f'[heartbeat] session timed out for agent "{agent.id}"')
            return _fail_safe_result(agent.id), client

        if not response_text:
            logger.error(f'[heartbeat] no response text for agent "{agent.id}"')
            return _fail_safe_result(agent.id), client

        result = _parse_heartbeat_response(agent.id, response_text)
        return result, client

    except Exception as e:
        if timed_out:
            logger.error(f'[heartbeat] session timed out for agent "{agent.id}"')
        else:
            logger.error(f'[heartbeat] session error for agent "{agent.id}": {e}')
        return _fail_safe_result(agent.id), client

    finally:
        timeout_task.cancel()


async def initiate_agent_call(agent: Agent, client: ClaudeSDKClient) -> str:
    """Initiate an outbound Twilio call for an agent.

    Stores the live client in pending_calls, places the outbound call via Twilio
    REST API, and sets a 60-second cleanup timer.

    Args:
        agent: Full agent data
        client: Live Claude session to hand off to the voice pipeline

    Returns:
        The Twilio call SID
    """
    if not _config:
        raise RuntimeError("Heartbeat not started -- no config")

    tunnel_url = _get_tunnel_url() if _get_tunnel_url else None
    if not tunnel_url:
        raise RuntimeError("Tunnel is not running. Cannot place outbound call.")

    if not _config.twilio_account_sid or not _config.twilio_auth_token:
        raise RuntimeError("TWILIO_ACCOUNT_SID and TWILIO_AUTH_TOKEN must be set")

    if not _config.user_phone_number:
        raise RuntimeError("USER_PHONE_NUMBER must be set in Settings > General")

    token = str(uuid4())

    # Store the pending call with the live client
    pending = PendingCall(
        token=token,
        agent_id=agent.id,
        client=client,
        initial_prompt="The user just answered your call. Greet them and briefly explain why you're calling.",
        created_at=time.time(),
    )
    _pending_calls[token] = pending

    # Schedule cleanup for unanswered calls
    asyncio.create_task(_cleanup_pending_call(token))

    # Get from number via Twilio API
    from twilio.rest import Client as TwilioClient

    twilio_client = TwilioClient(_config.twilio_account_sid, _config.twilio_auth_token)
    numbers = twilio_client.incoming_phone_numbers.list(limit=1)
    if not numbers:
        raise RuntimeError("No Twilio phone numbers found on the account")
    from_number = numbers[0].phone_number

    # Build TwiML with WebSocket stream URL
    tunnel_host = tunnel_url.replace("https://", "").replace("http://", "")
    twiml = (
        f'<Response><Connect>'
        f'<Stream url="wss://{tunnel_host}/media/{token}?agentId={agent.id}" />'
        f'</Connect></Response>'
    )

    call = twilio_client.calls.create(
        to=_config.user_phone_number,
        from_=from_number,
        twiml=twiml,
    )

    logger.info(
        f"[heartbeat] outbound call placed to {_config.user_phone_number} "
        f"(callSid={call.sid})"
    )
    return call.sid


async def _cleanup_pending_call(token: str) -> None:
    """Clean up a pending call after PENDING_CALL_TIMEOUT_S if not claimed.

    Args:
        token: The pending call token
    """
    await asyncio.sleep(PENDING_CALL_TIMEOUT_S)

    pending = _pending_calls.pop(token, None)
    if pending and pending.client:
        logger.info(f"[heartbeat] cleaning up unanswered pending call: {token}")
        try:
            await pending.client.disconnect()
        except Exception:
            pass


def _parse_heartbeat_response(agent_id: str, text: str) -> HeartbeatResult:
    """Parse a heartbeat JSON response string into a HeartbeatResult.

    Expects a JSON object with shouldCall (boolean) and reason (string).

    Args:
        agent_id: Agent identifier
        text: Raw text from the assistant response

    Returns:
        Parsed HeartbeatResult
    """
    try:
        match = re.search(r'\{[\s\S]*"shouldCall"[\s\S]*\}', text)
        if not match:
            logger.error(
                f'[heartbeat] no JSON found in response for agent "{agent_id}": {text}'
            )
            return _fail_safe_result(agent_id)

        parsed = json.loads(match.group(0))
        return HeartbeatResult(
            agent_id=agent_id,
            should_call=bool(parsed.get("shouldCall", False)),
            reason=str(parsed.get("reason", "")),
            timestamp=time.time(),
        )
    except (json.JSONDecodeError, Exception) as e:
        logger.error(f'[heartbeat] JSON parse error for agent "{agent_id}": {e}')
        return _fail_safe_result(agent_id)


def _fail_safe_result(agent_id: str) -> HeartbeatResult:
    """Create a fail-safe HeartbeatResult that does not trigger a call.

    Args:
        agent_id: Agent identifier

    Returns:
        HeartbeatResult with should_call=False
    """
    return HeartbeatResult(
        agent_id=agent_id,
        should_call=False,
        reason="heartbeat check failed or timed out",
        timestamp=time.time(),
    )
